cpp/src/utilities/system.cpp

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/utilities/system.hpp>
+
+#include <rmm/cuda_device.hpp>
+
+#include <cuda_runtime.h>
+
+#include <nvml.h>
+
+#include <iostream>
+#include <mutex>
+#include <sstream>
+
+#define CHECK_NVML(err_code) check_nvml(err_code, __FILE__, __LINE__)
+inline void check_nvml(nvmlReturn_t err_code, const char* file, int line)
+{
+  if (err_code != nvmlReturn_enum::NVML_SUCCESS) {
+    std::stringstream ss;
+    ss << "NVML error: " << err_code << " " << nvmlErrorString(err_code) << " in " << file
+       << " at line " << line << std::endl;
+    throw std::runtime_error(ss.str());
+  }
+}
+
+namespace cudf {
+namespace detail {
+/**
+ * @brief Perform the initialization of NVML library.
+ *
+ * NVML initialization can be costly. Using this function ensures it is performed only once per
+ * process.
+ *
+ * @note NVML shutdown is also costly. We may intentionally not perform shutdown at the expense of a
+ * slight NVML resource leak.
+ */
+void initialize_nvml()
+{
+  static std::once_flag flag{};
+  std::call_once(flag, [] { CHECK_NVML(nvmlInit_v2()); });
+}
+}  // namespace detail
+
+bool is_c2c_available()
+{
+  // todo: remove this once CUDA 11 support is dropped
+#if CUDART_VERSION < 12000
```

Comment:
These macros are nearly equivalent but `CUDA_VERSION` is slightly more appropriate since you're dealing with information from a library that ships with the driver, not the runtime.
```suggestion
#if CUDA_VERSION < 12000
```